File: drawing_tool_pages/utils/portrait_processor.py
"""
Portrait Processor: MediaPipe landmark extraction, 3D anatomical frame calculation,
and Loomis precision scale & wireframe cranial cage construction.
"""

import os
import urllib.request
import numpy as np
import trimesh
import cv2
from PIL import Image
from scipy.spatial import KDTree
from scipy.interpolate import RBFInterpolator
from . import landmark_indices

# Paths to model OBJ files
CANONICAL_OBJ_PATH = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..", "test", "canonical_face_model.obj")
)
CANONICAL_HEAD_OBJ_PATH = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..", "test", "canonical_head_model.obj")
)

MODEL_TASK_PATH = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "utils", "face_landmarker.task")
)
MODEL_TASK_URL = "https://storage.googleapis.com/mediapipe-models/face_landmarker/face_landmarker/float16/1/face_landmarker.task"


def ensure_model_asset() -> str | None:
    """Ensure the MediaPipe face_landmarker.task model file is present locally."""
    if os.path.exists(MODEL_TASK_PATH):
        return MODEL_TASK_PATH
    try:
        os.makedirs(os.path.dirname(MODEL_TASK_PATH), exist_ok=True)
        print("Downloading MediaPipe face_landmarker.task model...")
        urllib.request.urlretrieve(MODEL_TASK_URL, MODEL_TASK_PATH)
        return MODEL_TASK_PATH
    except Exception as e:
        print(f"Warning: Failed to download MediaPipe model: {e}")
        return None


def load_canonical_mesh() -> trimesh.Trimesh:
    """Load default canonical face model mesh."""
    if os.path.exists(CANONICAL_OBJ_PATH):
        return trimesh.load(CANONICAL_OBJ_PATH, force="mesh", process=False)
    else:
        return trimesh.creation.icosphere(subdivisions=3, radius=8.0)


def extract_landmarks_from_image(image_input) -> np.ndarray | None:
    """Extract 478 3D landmarks (x, y, z) from an input image using MediaPipe FaceLandmarker."""
    try:
        import mediapipe as mp
        
        model_file = ensure_model_asset()
        if not model_file:
            return None

        if isinstance(image_input, str):
            image = cv2.imread(image_input)
            if image is None:
                return None
            image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        elif isinstance(image_input, np.ndarray):
            image_rgb = image_input
        elif isinstance(image_input, Image.Image):
            image_rgb = np.array(image_input)
        else:
            return None

        h, w, _ = image_rgb.shape
        mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=image_rgb)
        
        BaseOptions = mp.tasks.BaseOptions
        FaceLandmarker = mp.tasks.vision.FaceLandmarker
        FaceLandmarkerOptions = mp.tasks.vision.FaceLandmarkerOptions
        VisionRunningMode = mp.tasks.vision.RunningMode

        options = FaceLandmarkerOptions(
            base_options=BaseOptions(model_asset_path=model_file),
            running_mode=VisionRunningMode.IMAGE,
            num_faces=1
        )

        with FaceLandmarker.create_from_options(options) as landmarker:
            result = landmarker.detect(mp_image)
            
            if not result.face_landmarks or len(result.face_landmarks) == 0:
                return None

            landmarks = result.face_landmarks[0]
            coords = []
            for lm in landmarks:
                x = (lm.x - 0.5) * w / max(h, w)
                y = -(lm.y - 0.5) * h / max(h, w)
                z = -lm.z * w / max(h, w)
                coords.append([x, y, z])
                
            return np.array(coords)
    except Exception as e:
        print(f"Error extracting landmarks: {e}")
        return None


def deform_face_mesh(base_mesh: trimesh.Trimesh, landmarks_3d: np.ndarray) -> trimesh.Trimesh:
    """Deform base canonical face mesh vertices smoothly to align with photo landmarks."""
    deformed_mesh = base_mesh.copy()
    verts = deformed_mesh.vertices.copy()
    
    num_verts = len(verts)
    num_lms = len(landmarks_3d)
    min_count = min(num_verts, num_lms)
    
    if min_count > 10:
        base_pts = verts[:min_count]
        target_pts = landmarks_3d[:min_count]
        
        target_center = np.mean(target_pts, axis=0)
        target_centered = target_pts - target_center
        
        base_center = np.mean(base_pts, axis=0)
        base_scale = np.max(np.linalg.norm(base_pts - base_center, axis=1))
        target_scale = np.max(np.linalg.norm(target_centered, axis=1))
        
        if target_scale > 1e-6:
            scaled_target = target_centered * (base_scale / target_scale) + base_center
            verts[:min_count] = 0.85 * scaled_target + 0.15 * base_pts
            
    deformed_mesh.vertices = verts
    return deformed_mesh


def compute_anatomical_frame(face_mesh: trimesh.Trimesh):
    """
    Calculate orthogonal anatomical coordinate frame (X_axis, Y_axis, Z_axis)
    from MediaPipe face landmarks.
    """
    verts = face_mesh.vertices
    glabella = verts[168]      # Brow line
    hairline = verts[10]       # Hairline
    subnasale = verts[2]       # Nose base
    chin = verts[152]          # Chin tip
    left_temple = verts[234]   # Right side of screen / subject left
    right_temple = verts[454]  # Left side of screen / subject right

    x_axis = right_temple - left_temple
    x_axis /= np.linalg.norm(x_axis)

    y_axis = hairline - chin
    y_axis = y_axis - np.dot(y_axis, x_axis) * x_axis
    y_axis /= np.linalg.norm(y_axis)

    z_axis = np.cross(x_axis, y_axis)
    z_axis /= np.linalg.norm(z_axis)

    return {
        "glabella": glabella,
        "hairline": hairline,
        "subnasale": subnasale,
        "chin": chin,
        "left_temple": left_temple,
        "right_temple": right_temple,
        "x_axis": x_axis,
        "y_axis": y_axis,
        "z_axis": z_axis,
    }


def build_wireframe_cranial_sphere_geometry(face_mesh: trimesh.Trimesh) -> trimesh.Trimesh:
    """Build transformed cranial sphere geometry for wireframe edge rendering."""
    frame = compute_anatomical_frame(face_mesh)
    glabella = frame["glabella"]
    hairline = frame["hairline"]
    chin = frame["chin"]
    x_axis = frame["x_axis"]
    y_axis = frame["y_axis"]
    z_axis = frame["z_axis"]

    face_h = np.linalg.norm(hairline - chin)
    radius = float(face_h * 0.50)
    center = glabella - (z_axis * (radius * 1.02)) + (y_axis * (radius * 0.04))

    sphere_mesh = trimesh.creation.icosphere(subdivisions=2, radius=radius)
    s_verts = sphere_mesh.vertices.copy()
    side_clip = radius * 0.84
    s_verts[:, 0] = np.clip(s_verts[:, 0], -side_clip, side_clip)
    sphere_mesh.vertices = s_verts

    rot_mat = np.eye(4)
    rot_mat[:3, 0] = x_axis
    rot_mat[:3, 1] = y_axis
    rot_mat[:3, 2] = z_axis
    sphere_mesh.apply_transform(rot_mat)
    sphere_mesh.apply_translation(center)
    return sphere_mesh


def compute_loomis_measurement_scale(face_mesh: trimesh.Trimesh):
    """
    Compute 3D Parallel Measurement Scale Rod and Ticks (Measured Anatomy vs Ideal Loomis Thirds).
    """
    frame = compute_anatomical_frame(face_mesh)
    glabella = frame["glabella"]
    hairline = frame["hairline"]
    subnasale = frame["subnasale"]
    chin = frame["chin"]
    left_temple = frame["left_temple"]
    right_temple = frame["right_temple"]
    x_axis = frame["x_axis"]
    y_axis = frame["y_axis"]
    z_axis = frame["z_axis"]

    face_h = np.linalg.norm(hairline - chin)

    y_hairline = np.dot(hairline, y_axis)
    y_brow = np.dot(glabella, y_axis)
    y_nose = np.dot(subnasale, y_axis)
    y_chin = np.dot(chin, y_axis)

    total_h = y_hairline - y_chin
    ideal_unit = total_h / 3.0
    y_ideal_brow = y_hairline - ideal_unit
    y_ideal_nose = y_hairline - (2.0 * ideal_unit)

    scale_x_offset = np.linalg.norm(right_temple - left_temple) * 0.42
    scale_z_offset = max(12.0, face_h * 0.15)
    scale_base_origin = chin + (x_axis * scale_x_offset) + (z_axis * scale_z_offset)

    bar_start = scale_base_origin + y_axis * (y_chin - np.dot(scale_base_origin, y_axis))
    bar_end = scale_base_origin + y_axis * (y_hairline - np.dot(scale_base_origin, y_axis))
    scale_bar_pts = np.array([bar_start, bar_end])

    tick_len = max(12.0, face_h * 0.15)

    def make_tick(y_val, length=tick_len):
        pt_base = scale_base_origin + y_axis * (y_val - np.dot(scale_base_origin, y_axis))
        pt_ext = pt_base + (x_axis * length)
        return np.array([pt_base, pt_ext])

    ticks_detected = [
        make_tick(y_hairline),
        make_tick(y_brow),
        make_tick(y_nose),
        make_tick(y_chin),
    ]

    ticks_ideal = [
        make_tick(y_ideal_brow, length=tick_len * 1.3),
        make_tick(y_ideal_nose, length=tick_len * 1.3),
    ]

    return {
        "scale_bar_pts": scale_bar_pts,
        "ticks_detected": ticks_detected,
        "ticks_ideal": ticks_ideal,
    }


def compute_construction_rings(face_mesh: trimesh.Trimesh):
    """
    Compute 3D Construction Rings: Brow Ring, Midline Ring, Coronal Ring, Side Discs & Crosshairs.
    """
    frame = compute_anatomical_frame(face_mesh)
    glabella = frame["glabella"]
    hairline = frame["hairline"]
    chin = frame["chin"]
    x_axis = frame["x_axis"]
    y_axis = frame["y_axis"]
    z_axis = frame["z_axis"]

    face_h = np.linalg.norm(hairline - chin)
    radius = float(face_h * 0.50)
    center = glabella - (z_axis * (radius * 1.02)) + (y_axis * (radius * 0.04))
    side_clip = radius * 0.84

    theta = np.linspace(0, 2 * np.pi, 96)
    
    brow_ring_pts = np.array([
        center + (radius * 0.99) * (0.86 * np.cos(t) * x_axis - np.sin(t) * z_axis)
        for t in theta
    ])
    
    midline_ring_pts = np.array([
        center + (radius * 0.99) * (np.cos(t) * y_axis - np.sin(t) * z_axis)
        for t in theta
    ])
    
    coronal_ring_pts = np.array([
        center + (radius * 0.99) * (0.86 * np.cos(t) * x_axis + np.sin(t) * y_axis)
        for t in theta
    ])

    disc_radius = radius * 0.53
    side_rings_pts = []
    side_cross_pts = []
    
    for side_sign in [-1.0, 1.0]:
        disc_center = center + (side_sign * side_clip * x_axis)
        ring_pts = np.array([
            disc_center + disc_radius * (np.cos(t) * z_axis + np.sin(t) * y_axis)
            for t in theta
        ])
        side_rings_pts.append(ring_pts)
        
        cross_h = np.array([disc_center - (disc_radius * z_axis), disc_center + (disc_radius * z_axis)])
        cross_v = np.array([disc_center - (disc_radius * y_axis), disc_center + (disc_radius * y_axis)])
        side_cross_pts.extend([cross_h, cross_v])

    return {
        "brow_ring_pts": brow_ring_pts,
        "midline_ring_pts": midline_ring_pts,
        "coronal_ring_pts": coronal_ring_pts,
        "side_rings_pts": side_rings_pts,
        "side_cross_pts": side_cross_pts,
    }

File: src/utils/portraitProcessor.ts
import * as THREE from 'three';
import { FilesetResolver, FaceLandmarker } from '@mediapipe/tasks-vision';

let landmarkerInstance: FaceLandmarker | null = null;

export function parseCanonicalObj(objText: string): {
  geometry: THREE.BufferGeometry;
  vertices: THREE.Vector3[];
} {
  const lines = objText.split('\n');
  const positions: number[] = [];
  const vertices: THREE.Vector3[] = [];
  const indices: number[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith('v ')) {
      const parts = trimmed.split(/\s+/).filter(Boolean).slice(1).map(Number);
      if (parts.length >= 3) {
        positions.push(parts[0], parts[1], parts[2]);
        vertices.push(new THREE.Vector3(parts[0], parts[1], parts[2]));
      }
    } else if (trimmed.startsWith('f ')) {
      const parts = trimmed.split(/\s+/).filter(Boolean).slice(1);
      const faceIndices = parts.map((p) => {
        const vIdx = parseInt(p.split('/')[0], 10);
        return vIdx - 1; // Convert 1-indexed OBJ to 0-indexed vertex array
      });
      if (faceIndices.length === 3) {
        indices.push(faceIndices[0], faceIndices[1], faceIndices[2]);
      } else if (faceIndices.length === 4) {
        indices.push(faceIndices[0], faceIndices[1], faceIndices[2]);
        indices.push(faceIndices[0], faceIndices[2], faceIndices[3]);
      }
    }
  }

  const geometry = new THREE.BufferGeometry();
  geometry.setAttribute('position', new THREE.Float32BufferAttribute(positions, 3));
  geometry.setIndex(indices);
  geometry.computeVertexNormals();

  return { geometry, vertices };
}

export async function initFaceLandmarker(basePath: string = ''): Promise<FaceLandmarker | null> {
  if (landmarkerInstance) return landmarkerInstance;
  try {
    const vision = await FilesetResolver.forVisionTasks(
      'https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@latest/wasm'
    );
    const modelPath = `${basePath}/models/face_landmarker.task`.replace(/\/+/g, '/');
    landmarkerInstance = await FaceLandmarker.createFromOptions(vision, {
      baseOptions: {
        modelAssetPath: modelPath,
        delegate: 'GPU',
      },
      runningMode: 'IMAGE',
      numFaces: 1,
    });
    return landmarkerInstance;
  } catch (err) {
    console.warn('GPU delegate fallback to CPU for MediaPipe FaceLandmarker:', err);
    try {
      const vision = await FilesetResolver.forVisionTasks(
        'https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@latest/wasm'
      );
      const modelPath = `${basePath}/models/face_landmarker.task`.replace(/\/+/g, '/');
      landmarkerInstance = await FaceLandmarker.createFromOptions(vision, {
        baseOptions: {
          modelAssetPath: modelPath,
          delegate: 'CPU',
        },
        runningMode: 'IMAGE',
        numFaces: 1,
      });
      return landmarkerInstance;
    } catch (e) {
      console.error('Failed to initialize FaceLandmarker:', e);
      return null;
    }
  }
}

export async function extractLandmarksFromImage(
  imageSource: HTMLImageElement | HTMLCanvasElement,
  basePath: string = ''
): Promise<THREE.Vector3[] | null> {
  const landmarker = await initFaceLandmarker(basePath);
  if (!landmarker) return null;

  try {
    const result = landmarker.detect(imageSource);
    if (!result.faceLandmarks || result.faceLandmarks.length === 0) {
      return null;
    }

    const landmarks = result.faceLandmarks[0];
    const w = imageSource.width || 500;
    const h = imageSource.height || 500;
    const maxDim = Math.max(h, w);

    const coords: THREE.Vector3[] = landmarks.map((lm) => {
      const x = ((lm.x - 0.5) * w) / maxDim;
      const y = -((lm.y - 0.5) * h) / maxDim;
      const z = (-lm.z * w) / maxDim;
      return new THREE.Vector3(x, y, z);
    });

    return coords;
  } catch (e) {
    console.error('Error extracting face landmarks:', e);
    return null;
  }
}

export interface AnatomicalFrame {
  glabella: THREE.Vector3;
  hairline: THREE.Vector3;
  subnasale: THREE.Vector3;
  chin: THREE.Vector3;
  leftTemple: THREE.Vector3;
  rightTemple: THREE.Vector3;
  xAxis: THREE.Vector3;
  yAxis: THREE.Vector3;
  zAxis: THREE.Vector3;
}

export function computeAnatomicalFrame(vertices: THREE.Vector3[]): AnatomicalFrame {
  const glabella = vertices[168] ? vertices[168].clone() : new THREE.Vector3(0, 2, 0);
  const hairline = vertices[10] ? vertices[10].clone() : new THREE.Vector3(0, 5, 0);
  const subnasale = vertices[2] ? vertices[2].clone() : new THREE.Vector3(0, -1, 0);
  const chin = vertices[152] ? vertices[152].clone() : new THREE.Vector3(0, -5, 0);
  const leftTemple = vertices[234] ? vertices[234].clone() : new THREE.Vector3(-3, 1, 0);
  const rightTemple = vertices[454] ? vertices[454].clone() : new THREE.Vector3(3, 1, 0);

  const xAxis = new THREE.Vector3().subVectors(rightTemple, leftTemple).normalize();

  const yAxis = new THREE.Vector3().subVectors(hairline, chin);
  const xProj = xAxis.clone().multiplyScalar(yAxis.dot(xAxis));
  yAxis.sub(xProj).normalize();

  const zAxis = new THREE.Vector3().crossVectors(xAxis, yAxis).normalize();

  return {
    glabella,
    hairline,
    subnasale,
    chin,
    leftTemple,
    rightTemple,
    xAxis,
    yAxis,
    zAxis,
  };
}

export function deformVertices(
  baseVertices: THREE.Vector3[],
  landmarks: THREE.Vector3[]
): THREE.Vector3[] {
  const minCount = Math.min(baseVertices.length, landmarks.length);
  if (minCount <= 10) return baseVertices.map((v) => v.clone());

  const deformed = baseVertices.map((v) => v.clone());

  const targetPts = landmarks.slice(0, minCount);
  const basePts = baseVertices.slice(0, minCount);

  const targetCenter = new THREE.Vector3();
  targetPts.forEach((p) => targetCenter.add(p));
  targetCenter.divideScalar(minCount);

  const baseCenter = new THREE.Vector3();
  basePts.forEach((p) => baseCenter.add(p));
  baseCenter.divideScalar(minCount);

  let baseScale = 0;
  basePts.forEach((p) => {
    baseScale = Math.max(baseScale, p.distanceTo(baseCenter));
  });

  let targetScale = 0;
  targetPts.forEach((p) => {
    targetScale = Math.max(targetScale, p.clone().sub(targetCenter).length());
  });

  if (targetScale > 1e-6) {
    const scaleFactor = baseScale / targetScale;
    for (let i = 0; i < minCount; i++) {
      const scaledTarget = targetPts[i]
        .clone()
        .sub(targetCenter)
        .multiplyScalar(scaleFactor)
        .add(baseCenter);
      deformed[i].lerpVectors(basePts[i], scaledTarget, 0.85);
    }
  }

  return deformed;
}

export function computeLoomisMeasurementScale(vertices: THREE.Vector3[]) {
  const frame = computeAnatomicalFrame(vertices);
  const { glabella, hairline, subnasale, chin, leftTemple, rightTemple, xAxis, yAxis, zAxis } = frame;

  const faceH = hairline.distanceTo(chin);

  const yHairline = hairline.dot(yAxis);
  const yBrow = glabella.dot(yAxis);
  const yNose = subnasale.dot(yAxis);
  const yChin = chin.dot(yAxis);

  const totalH = yHairline - yChin;
  const idealUnit = totalH / 3.0;
  const yIdealBrow = yHairline - idealUnit;
  const yIdealNose = yHairline - 2.0 * idealUnit;

  const scaleXOffset = rightTemple.distanceTo(leftTemple) * 0.42;
  const scaleZOffset = faceH * 0.15;

  const scaleBaseOrigin = chin
    .clone()
    .add(xAxis.clone().multiplyScalar(scaleXOffset))
    .add(zAxis.clone().multiplyScalar(scaleZOffset));

  const barStart = scaleBaseOrigin
    .clone()
    .add(yAxis.clone().multiplyScalar(yChin - scaleBaseOrigin.dot(yAxis)));
  const barEnd = scaleBaseOrigin
    .clone()
    .add(yAxis.clone().multiplyScalar(yHairline - scaleBaseOrigin.dot(yAxis)));

  const scaleBarPts: [THREE.Vector3, THREE.Vector3] = [barStart, barEnd];
  const tickLen = faceH * 0.15;

  const makeTick = (yVal: number, length: number = tickLen): [THREE.Vector3, THREE.Vector3] => {
    const ptBase = scaleBaseOrigin
      .clone()
      .add(yAxis.clone().multiplyScalar(yVal - scaleBaseOrigin.dot(yAxis)));
    const ptExt = ptBase.clone().add(xAxis.clone().multiplyScalar(length));
    return [ptBase, ptExt];
  };

  const ticksDetected = [
    makeTick(yHairline),
    makeTick(yBrow),
    makeTick(yNose),
    makeTick(yChin),
  ];

  const ticksIdeal = [
    makeTick(yIdealBrow, tickLen * 1.3),
    makeTick(yIdealNose, tickLen * 1.3),
  ];

  return {
    scaleBarPts,
    ticksDetected,
    ticksIdeal,
  };
}

export function computeConstructionRings(vertices: THREE.Vector3[]) {
  const frame = computeAnatomicalFrame(vertices);
  const { glabella, hairline, chin, xAxis, yAxis, zAxis } = frame;

  const faceH = hairline.distanceTo(chin);
  const radius = faceH * 0.5;

  const center = glabella
    .clone()
    .sub(zAxis.clone().multiplyScalar(radius * 1.02))
    .add(yAxis.clone().multiplyScalar(radius * 0.04));

  const sideClip = radius * 0.84;
  const numSteps = 96;
  const thetaStep = (Math.PI * 2) / numSteps;

  const browRingPts: THREE.Vector3[] = [];
  const midlineRingPts: THREE.Vector3[] = [];
  const coronalRingPts: THREE.Vector3[] = [];

  for (let i = 0; i <= numSteps; i++) {
    const t = i * thetaStep;

    // Brow Ring
    const browPt = center
      .clone()
      .add(
        xAxis
          .clone()
          .multiplyScalar(radius * 0.99 * 0.86 * Math.cos(t))
      )
      .sub(
        zAxis
          .clone()
          .multiplyScalar(radius * 0.99 * Math.sin(t))
      );
    browRingPts.push(browPt);

    // Midline Ring
    const midPt = center
      .clone()
      .add(
        yAxis
          .clone()
          .multiplyScalar(radius * 0.99 * Math.cos(t))
      )
      .sub(
        zAxis
          .clone()
          .multiplyScalar(radius * 0.99 * Math.sin(t))
      );
    midlineRingPts.push(midPt);

    // Coronal Ring
    const corPt = center
      .clone()
      .add(
        xAxis
          .clone()
          .multiplyScalar(radius * 0.99 * 0.86 * Math.cos(t))
      )
      .add(
        yAxis
          .clone()
          .multiplyScalar(radius * 0.99 * Math.sin(t))
      );
    coronalRingPts.push(corPt);
  }

  const discRadius = radius * 0.53;
  const sideRingsPts: THREE.Vector3[][] = [];
  const sideCrossPts: [THREE.Vector3, THREE.Vector3][] = [];

  [-1.0, 1.0].forEach((sideSign) => {
    const discCenter = center.clone().add(xAxis.clone().multiplyScalar(sideSign * sideClip));
    const ringPts: THREE.Vector3[] = [];
    for (let i = 0; i <= numSteps; i++) {
      const t = i * thetaStep;
      const pt = discCenter
        .clone()
        .add(zAxis.clone().multiplyScalar(discRadius * Math.cos(t)))
        .add(yAxis.clone().multiplyScalar(discRadius * Math.sin(t)));
      ringPts.push(pt);
    }
    sideRingsPts.push(ringPts);

    const crossH: [THREE.Vector3, THREE.Vector3] = [
      discCenter.clone().sub(zAxis.clone().multiplyScalar(discRadius)),
      discCenter.clone().add(zAxis.clone().multiplyScalar(discRadius)),
    ];
    const crossV: [THREE.Vector3, THREE.Vector3] = [
      discCenter.clone().sub(yAxis.clone().multiplyScalar(discRadius)),
      discCenter.clone().add(yAxis.clone().multiplyScalar(discRadius)),
    ];
    sideCrossPts.push(crossH, crossV);
  });

  return {
    browRingPts,
    midlineRingPts,
    coronalRingPts,
    sideRingsPts,
    sideCrossPts,
  };
}
